---
title: "Analyzing GEO Microarray Data"
output: html_notebook
---

[How-to-analyze-GEO-microarray-data](https://github.com/Lindseynicer/How-to-analyze-GEO-microarray-data)

```{r}
plot(cars)
```

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Ctrl+Alt+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.

## GEO
The Gene Expression Omnibus (GEO) is a data repository hosted by the National Center for Biotechnology Information (NCBI). NCBI contains all publicly available nucleotide and protein sequences.

## GEOquery

```{r}
getwd()
```

Install [R tools](https://cran.rstudio.com/bin/windows/Rtools/)

```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("GEOquery")
```
Load `Geoquery`

```{r}
library(GEOquery)
```
Load data

```{r}
my_id <- "GSE63477"
gse <- getGEO(my_id)
```


```{r}
length(gse)
```


```{r}
gse <- gse[[1]]
gse
```
Print sample information

```{r}
pData(gse)[1:2,]

```
Print gene annotation

```{r}
fData(gse)[1,]
```
Print expression data

```{r}
exprs(gse)[1,]
```
## Check normalization and scales used
To check if data is RMA (Robust Multiarray Average) normalized to remove low-expression gene. Use the 'summary' function and check if the data is log transformed, i.e., within (0-16) range. Do the RMA normalization or log transform otherwise.

Performing log2 transform, drawing boxplot. Data is normalized if distributions are similar.

```{r}
pData(gse)$data_processing[1]
```

```{r}
summary(exprs(gse))
```


```{r}
exprs(gse) <- log2(exprs(gse))

summary(exprs(gse))
```


```{r}
boxplot(exprs(gse),outline=F)
```

## Inspect the clinical variables
Sample name is stored in column of 'characteristics_ch1.1'. We can select and rename the column.

```{r}
sampleInfo <- pData(gse)
head(sampleInfo)
```


```{r}
table(sampleInfo$characteristics_ch1.1)
```


```{r}
library(dplyr)
sampleInfo <- select(sampleInfo, characteristics_ch1.1)

#rename
sampleInfo <- rename(sampleInfo, sample=characteristics_ch1.1)
head(sampleInfo)
```


```{r}
dim(sampleInfo)
```


```{r}
sampleInfo$sample
```


```{r}
library(stringr)
sampleInfo$group <- ""
for(i in 1:nrow(sampleInfo)){
  if(str_detect(sampleInfo$sample[i], "CTRL") && str_detect(sampleInfo$sample[i], "full"))
  {sampleInfo$group[i] <- "Conf"}
  
  if(str_detect(sampleInfo$sample[i], "CTRL") && str_detect(sampleInfo$sample[i], "dextran"))
  {sampleInfo$group[i] <- "Cond"}
  
  if(str_detect(sampleInfo$sample[i], "cabazitaxel") && str_detect(sampleInfo$sample[i], "full"))
  {sampleInfo$group[i] <- "cabazitaxelf"}
  
  if(str_detect(sampleInfo$sample[i], "cabazitaxel") && str_detect(sampleInfo$sample[i], "dextran"))
  {sampleInfo$group[i] <- "cabazitaxeld"}
  
  if(str_detect(sampleInfo$sample[i], "docetaxel") && str_detect(sampleInfo$sample[i], "full"))
  {sampleInfo$group[i] <- "docetaxelf"}
  
  if(str_detect(sampleInfo$sample[i], "docetaxel") && str_detect(sampleInfo$sample[i], "dextran"))
  {sampleInfo$group[i] <- "docetaxeld"}
}

sampleInfo 
```


```{r}
sampleInfo$serum <- ""
for(i in 1:nrow(sampleInfo)){
  if(str_detect(sampleInfo$sample[i], "dextran"))
  {sampleInfo$serum[i] <- "dextran"}
  
  if(str_detect(sampleInfo$sample[i], "full"))
  {sampleInfo$serum[i] <- "full_serum"}
 
}

sampleInfo <- sampleInfo[,-1]
sampleInfo
```
## Sample clustering and Principle Components Analysis


```{r}
install.packages("pheatmap")
```

```{r}
library(pheatmap)
## argument use="c" stops an error if there are any missing data points

corMatrix <- cor(exprs(gse),use="c")
pheatmap(corMatrix)   
```

Check if row and column names match the correlation matrix otherwise format match.

```{r}
rownames(sampleInfo)
```


```{r}
colnames(corMatrix)
```


```{r}
## If not, force the rownames to match the columns
#rownames(sampleInfo) <- colnames(corMatrix)

pheatmap(corMatrix, annotation_col= sampleInfo)
```

Another way is to use PCA. It requires the data to be transposed, so that the genelist in the column, and rownames in samples.



```{r}
if (!require("ggplot2", quietly = TRUE))
    install.packages("ggplot2")

if (!require("ggrepel", quietly = TRUE))
    install.packages("ggrepel")
```


```{r}
#make PCA
library(ggplot2)
library(ggrepel)
## MAKE SURE TO TRANSPOSE THE EXPRESSION MATRIX

pca <- prcomp(t(exprs(gse)))

## Join the PCs to the sample information
cbind(sampleInfo, pca$x) %>% 
  ggplot(aes(x = PC1, y=PC2, col=group, label=paste("",group))) + geom_point() + geom_text_repel()
```

## Differential expression analysis
We'll use the limma (Linear models for microarray) package to perform differential expressions. We need to tell limma what sample groups we want to compare. I choose sampleInfo$group. A design matrix will be created, this is a matrix of 0 and 1, one row for each sample and one column for each sample group. We'll rename column names.

Now, let’s check if the expression data contain any lowly-expressed genes, this will affect the quality of DE analysis. A big problem in doing statistical analysis like limma is the inference of type 1 statistical errors, also called false positive. One simple way to reduce the possibility for type 1 errors is to do fewer comparisons, by filtering the data. For example, we know that not all genes are expressed in all tissues and many genes will not be expressed in any sample. As a result, in DGE analysis, it makes sense to remove the genes that are likely not expressed at all.

It is quite subjective how one defines a gene being expressed, here, I follow the tutorial, to make the cut off at the median of the expression values, which means to consider around 50% of the genes will not be expressed. Keep those expressed genes if they are present in more than 2 samples.

We can see that around half of the genes are not qualified as an “expressed” gene here, which makes sense, bcoz our cut-off is the median value.

```{r}
if (!require("limma", quietly = TRUE))
    install.packages("limma")

library(limma)
```


```{r}
design <- model.matrix(~0 + sampleInfo$group)
design
```


```{r}
## the column names are a bit ugly, so we will rename
colnames(design) <- c("Cabazitaxeld","Cabazitaxelf","Cond","Conf","Docetaxeld","Docetaxelf")

design
```


```{r}
## calculate median expression level
cutoff <- median(exprs(gse))

## TRUE or FALSE for whether each gene is "expressed" in each sample
is_expressed <- exprs(gse) > cutoff

## Identify genes expressed in more than 2 samples
keep <- rowSums(is_expressed) > 3

## check how many genes are removed / retained.
table(keep)
```


```{r}
## subset to just those expressed genes
gse <- gse[keep,]
```

Here there is a little extra step to find out the outliers. This has to be done carefully so the filtered data won’t be too biased. We calculate ‘weights’ to define the reliability of each sample. The ‘arrayweights’ function will assign a score to each sample, with a value of 1 implying equal weight. Samples with score less than 1 are down-weighed, or else up-weighed.

```{r}
# coping with outliers
## calculate relative array weights
aw <- arrayWeights(exprs(gse),design)
aw
```
Now we have a design matrix, we need to estimate the coefficients. For this design, we will essentially average the replicate arrays for each sample level. In addition, we will calculate standard deviations for each gene, and the average intensity for the genes across all microarrays.

We are ready to tell limma which pairwise contrasts that we want to make. For this experiment, we are going to contrast treatment (there are two types of texane drugs) and control in each serum type. So there are 4 contrasts to specify.

To do the statistical comparisons, Limma uses Bayesian statistics to minimize type 1 error. The eBayes function performs the tests. To summarize the results of the statistical test, ‘topTable’ will adjust the p-values and return the top genes that meet the cutoffs that you supply as arguments; while ‘decideTests’ will make calls for DEGs by adjusting the p-values and applying a logFC cutoff similar to topTable.

```{r}
## Fitting the coefficients
fit <- lmFit(exprs(gse), design,
             weights = aw)

head(fit$coefficients)
```


```{r}
## Making comparisons between samples, can define multiple contrasts
contrasts <- makeContrasts(Docetaxeld - Cond, Cabazitaxeld - Cond, Docetaxelf - Conf, Cabazitaxelf - Conf, levels = design)

fit2 <- contrasts.fit(fit, contrasts)
fit2 <- eBayes(fit2)


topTable(fit2)
```


```{r}
topTable1 <- topTable(fit2, coef=1)
topTable2 <- topTable(fit2, coef=2)
topTable3 <- topTable(fit2, coef=3)
topTable4 <- topTable(fit2, coef=4)

#if we want to know how many genes are differentially expressed overall, we can use the decideTest function.
summary(decideTests(fit2))
```


```{r}
table(decideTests(fit2))
```
## Further visualization with gene annotation
Now we want to know the gene name associated with the gene ID. The annotation data can be retrieved with the ‘fData’ function. Let’s select the ID, GB_ACC, this is genbank accession ID. Add into fit2 table.

The “Volcano Plot” function is a common way of visualising the results of a DE analysis. The x axis shows the log-fold change and the y axis is some measure of statistical significance, which in this case is the log-odds, or “B” statistic. We can also change the color of those genes with p value cutoff more than 0.05, and fold change cut off more than 1.

```{r}
anno <- fData(gse)
head(anno)
```


```{r}
anno <- select(anno,ID,GB_ACC)
fit2$genes <- anno

topTable(fit2)
```
```{r}
## Create volcano plot
full_results1 <- topTable(fit2, coef=1, number=Inf)
library(ggplot2)
ggplot(full_results1,aes(x = logFC, y=B)) + geom_point()
```
```{r}
## change according to your needs
p_cutoff <- 0.05
fc_cutoff <- 1


full_results1 %>% 
  mutate(Significant = P.Value < p_cutoff, abs(logFC) > fc_cutoff ) %>% 
  ggplot(aes(x = logFC, y = B, col=Significant)) + geom_point()
```

## Further visualization of selected gene
I think at this point, we are quite clear about data structure of GSE data. It has an experiment data, pData; the expression data, exprs; and also annotation data, fData. And we have learned how to check the expression data, normalize them, and perform differential expression analysis.

Now, with the differential expression gene tables, there are some downstream analyses that we can continue, such as to export a full table of DE genes, to generate a heatmap for your selected genes, get the gene list for a particular pathway, or survival analysis (but this is only for those clinical data).

Here, I just want to look into the fold change data of a selected gene, whether it is significantly differential expressed or not.

```{r}
## Get the results for particular gene of interest
#GB_ACC for Nkx3-1 is NM_001256339 or NM_006167
##no NM_001256339 in this data
full_results2 <- topTable(fit2, coef=2, number=Inf)
full_results3 <- topTable(fit2, coef=3, number=Inf)
full_results4 <- topTable(fit2, coef=4, number=Inf)
filter(full_results1, GB_ACC == "NM_006167")
```

```{r}
filter(full_results2, GB_ACC == "NM_006167")
```

```{r}
filter(full_results3, GB_ACC == "NM_006167")
```

```{r}
filter(full_results4, GB_ACC == "NM_006167")
```
